### PR TITLE
pocl 3.0 (new formula)

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Experimenting with OpenCL formulae on Linux.

### CODE REVIEW

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:15]: Noting that I will wait for `llvm` v14 update to see if this needs to be downgraded.

File: Formula/ocl-icd.rb
Comment (Bo98) [reply Formula/ocl-icd.rb:1]: This was probably mostly because Kronos' version used to be nonfree. This is no longer the case.

I'm not sure what the functional differences are.

Comment (cho-m) [reply Formula/ocl-icd.rb:1]: Khronos' version does seem to be more actively maintained and may be better sync'd to latest changes.

It does seem `opencl-icd-loader`'s first release was "relatively" recent (`v2020.03.13`) while `ocd-icd` 1.0 was from 2012.

https://github.com/OCL-dev/ocl-icd/issues/11 mentions environment variables as main difference.

**Comment by cho-m** [comment]
Added both loader options: `ocl-icd` and `opencl-icd-loader`. Looking for feedback on which we want to use on Linux for `libOpenCL.so`.

Also can discuss other details on how we want to distribute OpenCL on Linux.

I am thinking that we can just leave the OpenCL runtime installation to user rather than requiring a dependency, but this means the dependents won't work out-of-the-box. However, it allows users to pick a hardware-specific runtime rather than having to install a common denominator like `pocl`. In many cases, a GPU runtime will be better like:
- Intel - https://github.com/intel/compute-runtime - Intel iGPUs from Broadwell onward
- AMD - https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime - for Radeon
- Nvidia - proprietary/non-free.

**Comment by cho-m** [comment]
Added both loader options: `ocl-icd` and `opencl-icd-loader`. Looking for feedback on which we want to use on Linux for `libOpenCL.so`.

Also can discuss other details on how we want to distribute OpenCL on Linux.

I am thinking that we can just leave the OpenCL runtime installation to user rather than requiring a dependency, but this means the dependents won't work out-of-the-box. However, it allows users to pick a hardware-specific runtime rather than having to install a common denominator like `pocl`. In many cases, a GPU runtime will be better like:
- Intel - https://github.com/intel/compute-runtime - Intel iGPUs from Broadwell onward
- AMD - https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime - for Radeon
- Nvidia - proprietary/non-free.

**Comment by danielnachun** [comment]
Just so I understand the plan, `opencl-headers` will be used to provide the headers for OpenCL applications, and for testing either `ocl-icd` or `opencl-icd-loader` will be used to load the CPU-based `pocl` OpenCL implementation.  Users will need to configure either `ocl-icd` or `opencl-icd-loader` to load their preferred OpenCL backend.  So this means that `ocl-icd` or `opencl-icd-loader` and `pocl` will be test dependencies on Linux and Apple Silicon (once it becomes ephemeral).  

I think we can provide both `ocl-icd` and `opencl-icd-loader` as options for users, but I think we should use `opencl-icd-loader` for testing because it seems to be more actively maintained.  

@cho-m do you consider this blocked by the LLVM 14 upgrade or do you think we can ship these as-is?  We have quite a few OpenCL formulae to fix on Linux I'd like to get this out soon.

**Comment by danielnachun** [comment]
> This won't fix ephemeral Apple Silicon as formulae need to directly link to Apple's OpenCL Framework. Based on information in PoCL, the Framework doesn't work with 3rd party ICD loaders http://portablecl.org/docs/html/using.html#using-pocl-on-macosx

That's good to know.  Hopefully Khronos or someone else will develop an OSS OpenCL wrapper for Metal similar to MoltenGL for OpenGL (though that is not OSS right now).

> Mainly looking for feedback for now before finalizing on approach so we don't have to backtrack/fixup later.
> 
> On Linux, some formulae may link to `libOpenCL.so` provided by the loader, which would make it a hard dependency. In many cases, it may just be easier to make ICD loader a hard dependency like Arch does, e.g. `hashcat` (https://archlinux.org/packages/community/x86_64/hashcat/) has dependency on `ocl-icd`. In this case, we probably want to settle on a since ICD loader and exclude the other one.
> PoCL 3.0 is now out which supports LLVM 14, so I'll use that in next run.

It sounds like making an ICD loader a hard dependency is the way to go then.  Let's use `opencl-icd-loader` because it is more actively maintained and is developed by Khronos.

**Comment by cho-m** [comment]
I think new PoCL 3.0 doesn't build with `opencl-icd-loader` due to https://github.com/pocl/pocl/commit/3ecda3b294d70e4f915ad141d669f4bc1298f606.

`OCL_ICD_INCLUDE_DIRS` only exists with `ocl-icd` as it looks for `ocl_icd.h` header from that implementation, which is generated by https://github.com/OCL-dev/ocl-icd/blob/v2.3.1/icd_generator.rb#L865.

**Comment by danielnachun** [comment]
In that case I guess the decision has been made for us.  I think we should still package `opencl-icd-loader` so that if the maintenance of `ocl-icd` becomes unsatisfactory, we'd have `opencl-icd-loader` ready to use.  Since there aren't that many OpenCL formulae, it wouldn't be hard to switch them over later if needed.  But for now let's see if we can get a working stack with `pocl` and `ocl-icd` and just treat `opencl-icd-loader` as a standalone formula.

Aside from rebasing/updating, is everything else ready to go for the formulae themselves?  If so I think we should be able to publish after another CI run.

**Comment by cho-m** [comment]
> Aside from rebasing/updating, is everything else ready to go for the formulae themselves? If so I think we should be able to publish after another CI run.

Still need to write tests for `opencl-icd-loader` and `pocl`.

I've split off `ocl-icd` to separate PR #104619 as that is probably ready to review and also the reason the PR is showing audit failure. I removed conflicts_with and will probably use keg_only for `opencl-icd-loader` as it is easier to switch between alternatives this way (based on previous experiences with other migrations).

There is a minor question on whether `etc` is best location. The ICD files need to be updated by corresponding formulae since they contain Cellar path. I think this does happen right now when file is installed by `cmake` as opposed to `etc.install`, though I don't know if that is intended.

**Comment by danielnachun** [comment]
I think this PR may need to be merged before https://github.com/Homebrew/homebrew-core/pull/104619 if we want to make `pocl` a dependency of `ocl-icd` (which I think we should do).  Since `pocl` isn't compatible with `opencl-icd-loader`, it shouldn't be a dependency for that and we should just make `opencl-icd-loader` available as an alternative for users who are using other OpenCL backends.

**Comment by cho-m** [comment]
> I think this PR may need to be merged before #104619 if we want to make `pocl` a dependency of `ocl-icd` (which I think we should do).

I don't think this is possible as it makes them depend on each other. `pocl` needs to find headers from `ocl-icd` to build with ICD support.

At best, we can add `pocl` as dependency to the OpenCL formulae that depend on `ocl-icd`.

**Comment by danielnachun** [comment]
Okay that makes sense, I didn't notice the build dependency.  I think it's fine to add `pocl` as a Linux-only dependency to all OpenCL packages.

**Comment by cho-m** [comment]
Split out `opencl-icd-loader` commits into separate PR #106164 as it is causing some issues when testing at same time as `pocl --> ocl-icd`.

Will drop commits from this PR.

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:22]: As discussed in `ocl-icd` PR, we will need to fix the ICD file before merge.

Could save into `#{libexec}/...` or `#{prefix}/etc/...` and then symlink into `#{etc}/OpenCL/vendors`

Comment (danielnachun) [reply Formula/pocl.rb:22]: Oh yeah I forgot about that.  I think that `#{prefix}/etc` makes more sense than `libexec` since it's just a text config file.

**Comment by danielnachun** [comment]
It looks like the `pocl` test fails on ARM.  Maybe this needs to be x86 only?

**Comment by cho-m** [comment]
> It looks like the `pocl` test fails on ARM. Maybe this needs to be x86 only?

Current failure is due to RPATH for libraries inside `#{lib}/pocl`. Using `-DCMAKE_INSTALL_RPATH=#{lib};#{lib}/pocl` should help (or maybe `-DCMAKE_INSTALL_RPATH=@loader_path;@loader_path/..` on macOS)
```console
❯ poclcc -l
LIST OF DEVICES:
0:
  Vendor:   ARM
    Name:   pthread
 Version:   OpenCL 3.0 PoCL HSTR: pthread-arm64-apple-macosx12.0.0-cyclone
```

However, fixing that leads to test failure from trying to use SSE2 rather than Cyclone:
```
Kernel library file /opt/homebrew/Cellar/pocl/3.0/lib/../share/pocl/kernel-arm64-apple-macosx12.0.0-sse2.bc doesn't exist.
```

**Comment by cho-m** [comment]
Looks like new error is due to `KERNELLIB_HOST_CPU_VARIANTS=distro` which is only supported on x86_64.

May need to check if this needs `ENV.runtime_cpu_detection` on x86_64 since `distro` is supposed to build for multiple AVX/SSE options.

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:21]: Still checking on this. Could alternatively use absolute paths and let brew handle rewrite, but I think we prefer paths that are already relocatable.
```suggestion
    # Add RPATH for libraries installed into #{lib}/pocl directory
    args << "-DCMAKE_INSTALL_RPATH=#{lib};#{lib}/pocl" if OS.mac?
```

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:23]: Also still need to confirm we can avoid `ENV.runtime_cpu_detection`. As mentioned, I think bit-code compilation is done outside our compiler shims as it uses LLVM directly.

**Comment by cho-m** [comment]
Also may need to check if LLVM Cellar will cause any issues:
```
/opt/homebrew/Cellar/pocl/3.0/lib/libpocl.2.9.0.dylib
--> match '/opt/homebrew/Cellar/llvm/14.0.6_1/bin/clang' at offset 0x7b111
```

File: Formula/pocl.rb
Comment (cho-m) [reply Formula/pocl.rb:21]: Looks like needs help on Linux too, so might just use absolute path rather than switching between `@loader_path` and `$ORIGIN` logic.
```
Full linkage --test pocl output
  Missing libraries:
    unexpected (libpocl-devices-basic.so)
```

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:28]: It looks like `#{prefix}/etc` is automatically symlinked to `#{etc}` via the Cellar path. Not sure how Homebrew handles this on updates (i.e. is there a broken symlink to Cellar path of previous version?)

I did try switching to `libexec` + `opt_libexec` in `install_symlink`, but it still uses the real Cellar path which is due to how `install_symlink` is implemented.

Looks like this needs a manual `ln_s` to use `opt...` paths.

Comment (cho-m) [reply Formula/pocl.rb:28]: Based on some quick observations after a few more attempts:
- `#{prefix}/etc` is automatically symlinked to `#{etc}`
- Manually symlinking `#{opt_prefix}/etc` to `#{etc}` causes `brew link` to fail due to above.
- Manually symlinking `#{prefix}/etc` to `#{etc}` is skipped.
- `(etc/...).install_symlink` and related logic (like above automatic symlink) seem to overwrite files on updates. Although this may work right now, I didn't see any documentation on this and behavior doesn't match `(etc/...).install`, so wasn't sure how reliable.
- Retrying `libexec` + `opt_libexec` did seem to work.

File: Formula/pocl.rb
Comment (cho-m) [inline Formula/pocl.rb:41]: Brew DSL doesn't handle `etc` well in this scenario as seen by various issues I mentioned in previous comment. Particularly, `etc.install_symlink` isn't well defined and behaves inconsistently.

I am now considering either manually creating ICD file or manually creating opt path symlink. 

```suggestion
    # The default ICD file contains the realpath to shared library, which breaks if
    # not overwritten. So, we cannot directly install generated ICD file into etc.
    # Symlinking files into etc exhibit unreliable behavior as it currently
    # overwrites files, which doesn't align with how etc.install behaves. Also,
    # manually running etc.install_symlink can produce broken symlinks on updates.
    # So, we manually create ICD file to fixed opt path with unversioned library.
    (buildpath/"pocl.icd").write opt_lib/shared_library("libpocl")
    (etc/"OpenCL/vendors").install "pocl.icd"
```

```suggestion
    # The default ICD file contains the realpath to shared library, which breaks if
    # not overwritten. So, we cannot directly install generated ICD file into etc.
    # etc.install_symlink into exhibits unreliable behavior as it currently
    # overwrites files, which doesn't align with how etc.install behaves. Also,
    # manually running etc.install_symlink can produce broken symlinks on updates.
    # So, we manually create symlink to ICD file if it doesn't exist.
    (etc/"OpenCL/vendors").mkpath
    icd = etc/"OpenCL/vendors/pocl.icd"
    ln_s opt_libexec/"OpenCL/vendors/pocl.icd", icd unless icd.exist?
```

**Review by danielnachun — APPROVED** [review]
Great job digging into the symlinking issue and fixing the build on ARM.  I think this is ready to go!

**Pull request closed without merging**
